Add Unicode Input Support for International Text in Maestro

salemaljebaly opened the pull request:
## Summary                                                               
                                                                                                                                                                     
This PR implements comprehensive Unicode input support for Maestro, addressing issue [#146](https://github.com/mobile-dev-inc/Maestro/issues/146). The implementation enables testing applications with international characters, emojis, RTL languages, and complex scripts through a new `maestro-unicode` module.                                                                      

https://github.com/user-attachments/assets/4ec72310-e2fc-4626-8b12-80daf94ac3e0


### Key Features                                                                                                                                                    
                                                                                                                                                                     
 - **Comprehensive Unicode Support**: Handles Latin extended, Arabic (RTL), Chinese (CJK), Japanese, Korean, and emoji characters                                    
 - **Smart Input Routing**: ASCII fast-path for performance, Unicode detection for complex characters                                                                
 - **Multiple Input Methods**: ADB keyboard integration with fallback strategies                                                                                     
 - **Configurable**: Extensive configuration options for timeouts, delays, and input methods                                                                         
                                                                                                                                                                     
 ### Architecture                                                                                                                                                    
                                                                                                                                                                     
 The implementation consists of several core components:                                                                                                             
                                                                                                                                                                     
 #### 1. UnicodeDetector (`maestro-unicode/src/main/kotlin/maestro/unicode/UnicodeDetector.kt`)                                                                      
 - Analyzes text to determine character sets (Latin, CJK, Arabic, Emoji, etc.)                                                                                       
 - Recommends optimal input methods based on text complexity                                                                                                         
                                                                                                                                                                     
 #### 2. UnicodeInputHandler (`maestro-unicode/src/main/kotlin/maestro/unicode/UnicodeInputHandler.kt`)                                                              
 - Main orchestration component for Unicode text input                                                                                                               
 - Handles asynchronous input with proper error handling and fallbacks                                                                                               
                                                                                                                                                                     
 #### 3. AdbKeyboardManager (`maestro-unicode/src/main/kotlin/maestro/unicode/AdbKeyboardManager.kt`)                                                                
 - Manages ADB Keyboard installation and lifecycle                                                                                                                   
 - Handles secure text transmission via Base64 encoding                                                                                                              
                                                                                                                                                                     
 #### 4. UnicodeConfig (`maestro-unicode/src/main/kotlin/maestro/unicode/config/UnicodeConfig.kt`)                                                                   
 - Centralized configuration management with validation                                                                                                              
 - Supports YAML-based configuration and preset configurations                                                                                                       
 - Includes timeout settings, delay configurations, and performance tuning options                                                                                   
                                                                                                                                                                     
 ### Integration Points                                                                                                                                              
                                                                                                                                                                     
 #### AndroidDriver Integration (`maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt`)                                                                    
 - Added `isUnicodeInputSupported()` method returning `true`                                                                                                         
 - Integrated Unicode input handling in `inputText()` method                                                                                                         
                                                                                                                                                                     
 #### Build Configuration                                                                                                                                            
 - Added `maestro-unicode` module to project structure                                                                                                               
 - Updated `settings.gradle.kts` and build dependencies                                                                                                              
 - Included ADB Keyboard APK as embedded resource                                                                                                                    
                                                                                                                                                                     
 ### Testing                                                                                                                                                         
                                                                                                                                                                     
 #### Unit Tests                                                                                                                                                     
 - **UnicodeDetectorTest**: Character set detection testing                                                                                            
 - **UnicodeConfigTest**: Configuration validation and loading tests                                                                                                 
 - Coverage includes edge cases, validation, and error handling                                                                                                      
                                                                                                                                                                     
 #### Integration Tests                                                                                                                                              
 - **Updated IntegrationTest**: Modified Case 037 to test Unicode input functionality                                                                                
 - **Enhanced Test Data**: Added comprehensive Unicode test cases including:                                                                                         
   - Latin extended characters (`Tést inpüt`)                                                                                                                        
   - Arabic RTL text (`مرحبا بالعالم`)                                                                                                                               
   - Chinese characters (`你好世界`)                                                                                                                                 
   - Japanese text (`こんにちは世界`)                                                                                                                                
   - Korean text (`안녕하세요`)                                                                                                                                      
   - Emoji support (`Hello 👋 World 🌍`)                                                                                                                             
   - Mixed language text (`Mixed مرحبا 你好`)                                                                                                                        
                                                                                                                                                                     
 #### FakeDriver Updates                                                                                                                                             
 - Enabled Unicode support in test driver for comprehensive testing                                                                                                  
 - Updated `isUnicodeInputSupported()` to return `true`                                                                                                              
                                                                                                                                                                     
 ### Files Changed                                                                                                                                                   
                                                                                                                                                                     
 #### New Files (1,712 additions)                                                                                                                                    
 - `maestro-unicode/build.gradle.kts` - Module build configuration                                                                                                   
 - `maestro-unicode/src/main/kotlin/maestro/unicode/UnicodeDetector.kt` - Character detection                                                                        
 - `maestro-unicode/src/main/kotlin/maestro/unicode/UnicodeInputHandler.kt` - Input orchestration                                                                    
 - `maestro-unicode/src/main/kotlin/maestro/unicode/AdbKeyboardManager.kt` - Keyboard management                                                                     
 - `maestro-unicode/src/main/kotlin/maestro/unicode/config/UnicodeConfig.kt` - Configuration                                                                         
 - `maestro-unicode/src/main/resources/ADBKeyboard.apk` - [Embedded keyboard APK](https://github.com/senzhk/ADBKeyBoard)                                                                                      
 - `maestro-unicode/src/test/kotlin/UnicodeDetectorTest.kt` - Unit tests                                                                                             
 - `maestro-unicode/src/test/kotlin/UnicodeConfigTest.kt` - Config tests                                                                                             
                                                                                                                                                                     
 #### Modified Files                                                                                                                                                 
 - `maestro-client/src/main/java/maestro/drivers/AndroidDriver.kt` - Unicode integration                                                                             
 - `maestro-test/src/main/kotlin/maestro/test/drivers/FakeDriver.kt` - Test support                                                                                  
 - `maestro-test/src/test/kotlin/maestro/test/IntegrationTest.kt` - Integration tests                                                                                
 - `maestro-test/src/test/resources/037_unicode_input.yaml` - Test data                                                                                              
 - `maestro-client/build.gradle.kts` - Dependencies                                                                                                                  
 - `settings.gradle.kts` - Module inclusion                                                                                                                          
                                                                                                                                                                     
 ### Performance Considerations                                                                                                                                      
                                                                                                                                                                     
 - **ASCII Fast-Path**: Maintains existing performance for ASCII-only text                                                                                           
 - **Text Chunking**: Handles large texts efficiently to avoid broadcast limits                                                                                      
 - **Keyboard Caching**: Reduces installation overhead through state caching                                                                                     
                                                                                                                                                                     
 ### Security & Privacy                                                                                                                                              
                                                                                                                                                                     
 - **APK Installation**: Uses embedded ADB Keyboard with system permissions                                                                                          
 - **Text Transmission**: Secure Base64 encoding for text data                                                                                                       
 - **Cleanup Procedures**: Proper resource cleanup and state management                                                                                              
                                                                                                                                                                     
 ## Test Plan                                                                                                                                                                                                   
                                                                                                                                                                     
 ### Automated Testing                                                                                                                                               
 - [x] All existing tests pass                                                                                                                                       
 - [x] New unit tests for Unicode components pass                                                                                                                    
 - [x] Integration test validates Unicode input functionality                                                                                                        
 - [x] Build and compilation successful                                                                                                                              
                                                                                                                                                                     
 ### Verification Commands                                                                                                                                           
                                                                                                                                                                     
 ```bash                                                                                                                                                             
 # Run all tests                                                                                                                                                     
 ./gradlew test                                                                                                                                                      
                                                                                                                                                                     
 # Run Unicode-specific tests                                                                                                                                        
 ./gradlew :maestro-unicode:test                                                                                                                                     
                                                                                                                                                                     
 # Run integration tests                                                                                                                                             
 ./gradlew :maestro-test:test                                                                                                                                        
                                                                                                                                                                     
 # Build the project                                                                                                                                                 
 ./gradlew build                                                                                                                                                     
 ```                                                                                                                                                                 
                                                                                                                                                                     
 ## Breaking Changes                                                                                                                                                 
                                                                                                                                                                     
 None. This implementation maintains full backward compatibility with existing ASCII input functionality.                                                            
                                                                                                                                                                     
 ## Future Enhancements                                                                                                                                              
                                                                                                                                                                                                                                           
 1. **Alternative Input Methods**: Reduce dependency on third-party keyboard                                                                                         
 2. **Expanded Language Support**: Add support for additional Unicode ranges and scripts                                                                             
                                                                                                                                                                     
 ## Related Issues                                                                                                                                                   
                                                                                                                                                                     
 Opened Issue [#146 ](https://github.com/mobile-dev-inc/Maestro/issues/146)- Add support for Unicode text input in mobile testing

salemaljebaly committed:
Add maestro-unicode module for Unicode text input support

  Introduces a new module that enables Unicode text input on Android devices
  through ADB Keyboard integration, supporting multiple writing systems and
  complex scripts.

  Core components:
  - UnicodeDetector: Analyzes text character sets and identifies Unicode content
  - UnicodeConfig: Configuration management with YAML support and defaults
  - AdbKeyboardManager: Handles ADB Keyboard installation and operation
  - UnicodeInputHandler: High-level Unicode input coordination
  - ADBKeyboard.apk: Embedded keyboard resource for automatic installation

  Features:
  - Support for RTL languages (Arabic, Hebrew)
  - Complex scripts (Thai, Devanagari, CJK characters)
  - Emoji and special symbols
  - Automatic fallback to standard input for ASCII text
  - Configurable performance settings

salemaljebaly committed:
Integrate Unicode support in AndroidDriver and build configuration

  Add Unicode text input capability to AndroidDriver using the maestro-unicode
  module. When Unicode characters are detected, the driver automatically switches
  to ADB Keyboard for proper input handling.

  Changes:
  - Add maestro-unicode dependency to maestro-client build configuration
  - Integrate Unicode detection and input handling in AndroidDriver.inputText()
  - Add isUnicodeInputSupported() method returning UnicodeConfig.isEnabled
  - Include maestro-unicode module in project settings
  - Maintain backward compatibility with ASCII text using standard input

salemaljebaly committed:
Add unit tests for Unicode module components

  Add comprehensive test coverage for the core Unicode detection and
  configuration functionality.

  Tests:
  - UnicodeDetectorTest: Validates character set detection, ASCII detection,
    bidirectional text detection, emoji detection, and input method recommendations
  - UnicodeConfigTest: Validates configuration loading from YAML, default values,
    customization options, and configuration validation methods

  Ensures reliable Unicode functionality with edge case coverage and
  validates integration with different character sets and writing systems.

salemaljebaly committed:
Update Unicode input test to support enabled Unicode functionality

- Enable Unicode support in FakeDriver for testing
- Update integration test to verify Unicode input works instead of throwing exception
- Expand test YAML with comprehensive Unicode character sets including Arabic, Chinese, Japanese, Korean, and emoji support

salemaljebaly committed:
Fix critical issues in Unicode input implementation

  - Fix emoji detection by properly handling Unicode surrogate pairs using codePointAt()
  - Fix thread safety in UnicodeConfig singleton with synchronized blocks and @Volatile
  - Fix resource leak in AdbKeyboardManager by using .use{} for APK stream handling
  - Update tests to reflect corrected emoji detection behavior

salemaljebaly commented:
Hi @herval 
E2e tests failing due to missing test apps (com.example.example, org.wikipedia).
I only added new Unicode module - no existing test/core code touched. 
Alos all tests pass locally with Java 17. Known flaky tests or missing CI setup?

Fishbowler commented on maestro-unicode/src/main/resources/ADBKeyboard.apk:
What's the source for this?

salemaljebaly replied:
The ADBKeyboard.apk is sourced from https://github.com/senzhk/ADBKeyBoard (GPL-2.0 license). It's required for Unicode input on Android since standard keyboards have limitations with complex Unicode characters that ADB input commands cannot handle.

Fishbowler replied:
I'm not a lawyer, but I don't think we can include GPL-2.0 binaries in this repo and be compliant

Fishbowler replied:
And that's nothing to say of whether we should be including binary blobs at all

salemaljebaly replied:
@herval @Fishbowler 
you’re absolutely right on both counts.

I’ve removed the ADBKeyboard.apk from the repository to avoid any GPL-2.0 licensing conflicts, as well as concerns around including precompiled binaries.

salemaljebaly replied:
Instead, I’ll update the implementation to download the APK at runtime directly from the official source (https://github.com/senzhk/ADBKeyBoard) when needed. This ensures we remain compliant with licensing requirements and maintain the security and transparency standards of the project.

droidchef commented:
Why can't we have just add the functionality that `ADBKeyboard` offers embedded in Maestro itself? 

AdbKeyboard has just one class in their service directory:
https://github.com/senzhk/ADBKeyBoard/blob/master/keyboardservice/src/main/java/com/android/adbkeyboard/AdbIME.java

No need to install APKs or Binaries you don't own or control

salemaljebaly committed:
Implement runtime APK download and improve Unicode input handling

  - Remove ADBKeyboard.apk binary from repository to resolve GPL-2.0 licensing concerns
  - Add runtime download functionality in AdbKeyboardManager from GitHub source
  - Fix priority order to attempt ADB Keyboard download first for Unicode text
  - Implement proper sendSingleCharacter method instead of placeholder
  - Fix clipboard fallback method with proper shell command syntax
  - Ensure automatic cleanup of temporary APK files after installation
  - Maintain full Unicode support functionality while addressing security concerns

salemaljebaly committed:
Add README documentation for maestro-unicode module

  - Document Unicode support features and capabilities
  - Explain runtime APK download process and cleanup behavior
  - Provide usage examples with various Unicode languages
  - Include configuration options and testing instructions
  - Address licensing and security concerns with clear explanations

Leland-Takamine commented:
> Why can't we have just add the functionality that `ADBKeyboard` offers embedded in Maestro itself?
> 
> AdbKeyboard has just one class in their service directory: https://github.com/senzhk/ADBKeyBoard/blob/master/keyboardservice/src/main/java/com/android/adbkeyboard/AdbIME.java
> 
> No need to install APKs or Binaries you don't own or control

If this is possible that indeed would be a great solution!

droidchef commented:
> > Why can't we have just add the functionality that `ADBKeyboard` offers embedded in Maestro itself?
> > AdbKeyboard has just one class in their service directory: https://github.com/senzhk/ADBKeyBoard/blob/master/keyboardservice/src/main/java/com/android/adbkeyboard/AdbIME.java
> > No need to install APKs or Binaries you don't own or control
> 
> @droidchef thank for replying
> 
> Based on my knowledge, Android requires Unicode input to go through an installed input method (IME) app. There’s no reliable way to bypass that using just ADB or Maestro — it’s a platform-level limitation.

As Leland mentioned, why don't we build the Input Method within Maestro and install that, rather than a 3P APK. 
Which is what I am supporting.

salemaljebaly commented:
> Thanks for quick reply @salemaljebaly. Downloading the binary would also pose a security concern from our perspective.
> 
> Here's two viable paths forward that we see for this:
> 
> 1. ADBKeyboard is actually a tiny amount of code. We can pretty easily reimplement the functionality in Maestro from the ground up (without taking any actual code from the original implementation as that would violate the license). And have our build process build the APK from Maestro source and bundle into the CLI.
> 2. Prompt the user when they attempt to input unicode on Android, telling them to install ADBKeyboard themselves. If they already have it installed, inputting unicode just works.
> 
> We prefer (1) since it solves the security and license issues while still remaining frictionless for the user. wdyt?

Thanks @Leland-Takamine  for the clear guidance and outlining the options.

You’re right — while runtime download avoids GPL issues (per the GPL FAQ), I fully understand and support the decision to avoid it for security reasons.

I’d like to move forward with Option 1 — reimplementing the functionality from scratch. This addresses the licensing and security concerns, keeps the UX smooth, and gives us full control.

The original functionality is minimal (handling a broadcast and inputting text), and I want to try building it myself — though it may take some time.

Would you prefer:
	1.	A separate maestro-keyboard module (APK)?
	2.	Integration into the main build?
	3.	Any architectural preferences?
